lib/src/css_mirroring/bidi_css_gen.dart

```diff
@@ -0,0 +1,334 @@
+///   If all the rulesets have to be removed -> remove the Directive itself.
+///
+/// If it is a Direction Independent Directive
+///   eg:
+///     @charset "UTF-8";                                 /// Charset Directive
+///     @namespace url(http://www.w3.org/1999/xhtml);     /// Namespace Directive
+///  Keep it in one of the transaction and remove it from other two (Here we are keeping it in Orientation Neutral transaction).
+///
+/// We then combine these transactions to get the expected output css.
+
+typedef Future<String> CssFlipper(String src);
+
+class _PendingRemovals {
+  final String source;
+  final TextEditTransaction _transaction;
+  // List to contain start and end location of pending removals.
+  var _startLocations = <int>[];
+  var _endLocations = <int>[];
+
+  _PendingRemovals(TextEditTransaction trans)
+      : _transaction = trans,
+        source = trans.file.getText(0);
+
+  void addRemoval(int start, int end) {
+    _startLocations.add(start);
+    _endLocations.add(end);
+  }
+
+  void commitRemovals() {
+    for (int iDecl = 0; iDecl < _startLocations.length; iDecl++) {
+      _transaction.edit(_startLocations[iDecl], _endLocations[iDecl], '');
+    }
+    _startLocations.clear();
+    _endLocations.clear();
+  }
+}
+
+class BidiCssGenerator {
+  final String _originalCss;
+  final String _flippedCss;
+  final String _cssSourceId;
+  final List<TreeNode> _originalTopLevels;
+  final List<TreeNode> _flippedTopLevels;
+  final Direction _nativeDirection;
+
+  BidiCssGenerator._(String originalCss, String flippedCss, this._cssSourceId, this._nativeDirection)
+      : _originalCss = originalCss,
+        _flippedCss = flippedCss,
+        _originalTopLevels = parse(originalCss).topLevels,
+        _flippedTopLevels = parse(flippedCss).topLevels;
+
+  static build(
+      String originalCss, String cssSourceId, Direction nativeDirection, CssFlipper cssFlipper) async {
+    return new BidiCssGenerator._(originalCss, await cssFlipper(originalCss), cssSourceId, nativeDirection);
+  }
+
+
+  /// main function which returns the bidirectional css.
+  String getOutputCss() {
+    var orientationNeutral = _makeTransaction(_originalCss, _cssSourceId);
+    var orientationSpecific = _makeTransaction(_originalCss, _cssSourceId);
+    var flippedOrientationSpecific = _makeTransaction(_flippedCss, _cssSourceId);
+    /// Modifies the transactions to contain only the desired parts.
+
+    _editTransaction(orientationNeutral, RetentionMode.keepBidiNeutral, _nativeDirection);
+    _editTransaction(orientationSpecific, RetentionMode.keepOriginalBidiSpecific, _nativeDirection);
+    _editTransaction(flippedOrientationSpecific, RetentionMode.keepFlippedBidiSpecific, flipDirection(_nativeDirection));
+
+    String getText(TextEditTransaction t) => (t.commit()..build('')).text;
+
+    return [
+      getText(orientationNeutral),
+      getText(orientationSpecific),
+      getText(flippedOrientationSpecific)
+    ].join('\n');
+  }
```

Comment:
Let's go further:

```
  String _buildCssFragment(String css, RetentionMode mode, Direction dir) {
    var tr = _makeTransaction(css, _cssSourceId);
    _editTransaction(tr, mode, dir);
    return (t.commit()..build('')).text;
  }

  String getOutputCss() => [
    _buildCssFragment(_originalCss, RetentionMode.keepBidiNeutral, _nativeDirection),
    _buildCssFragment(_originalCss, RetentionMode.keepOriginalBidiSpecific, _nativeDirection),
    _buildCssFragment(_flippedCss, RetentionMode.keepFlippedBidiSpecific, flipDirection(_nativeDirection))
  ].join('\n');
```
(can even fold _editTransaction into _buildCssFragment into, as the former's name is still problematic anyway)